marin/processing/tokenize/download_pretokenized.py

```diff
@@ -0,0 +1,146 @@
+"""
+Uses download_hf to download a pretokenized dataset cache from Hugging Face
+and prepares it as a tokenized dataset source for Levanter.
+"""
+
+import dataclasses
+import logging
+
+from levanter.data.text import (
+    LmDatasetFormatBase,
+    LMDatasetSourceConfig,
+    TextLmDatasetFormat,
+    UrlDatasetSourceConfig,
+)
+from levanter.store.cache import CacheOptions
+
+from marin.execution import THIS_OUTPUT_PATH, ExecutorStep, InputName
+from marin.processing.tokenize.tokenize import TokenizeConfigBase
+from operations.download.huggingface.download import DownloadConfig as HfDownloadConfig
+from operations.download.huggingface.download_hf import (
+    download_hf as hf_download_logic,
+)
+
+logger = logging.getLogger(__name__)
+
+
+@dataclasses.dataclass
+class PretokenizedCacheDownloadConfig(TokenizeConfigBase):
+    """Configuration for downloading a pre-existing Levanter cache from Hugging Face."""
+
+    # Fields from TokenizeConfigBase
+    cache_path: str  # Resolved by Executor: where the cache will be downloaded to
+    tokenizer: str  # Tokenizer name/path associated with this cache
+
+    # Fields specific to downloading from HF
+    hf_repo_id: str  # Hugging Face repository ID (e.g., "username/my_cache_repo")
+    hf_revision: str | None = None  # Revision, branch, or tag on Hugging Face
+    hf_repo_type_prefix: str = "datasets"  # Typically "datasets" or "models"
+    hf_token: str | None = None  # Hugging Face API token for private repositories
+
+    format: LmDatasetFormatBase = TextLmDatasetFormat()  # noqa: RUF009
+    cache_options: CacheOptions | None = None  # For TokenizeConfigBase interface, not used during download
+
+    tags: list[str] = dataclasses.field(default_factory=list)  # Tags for Levanter's LMDatasetSourceConfig
+
+    def as_lm_dataset_source_config(
+        self, actual_output_path: str | InputName | None, *, include_raw_paths=True
+    ) -> LMDatasetSourceConfig:
+        """
+        Returns a Levanter LMDatasetSourceConfig that points to the downloaded cache.
+        Since the cache is already tokenized and in Levanter format, train_urls and validation_urls
+        are empty. Levanter will load directly from the cache_dir.
+        """
+        if actual_output_path is None:
+            raise ValueError("actual_output_path must be provided for a downloaded cache.")
+
+        return UrlDatasetSourceConfig(
+            tags=self.tags,
+            train_urls=["dummy.jsonl"],  # Levanter misbehaves if these are empty
```

Comment:
We should fix this in levanter?  Put TODO or something?

Reply:
already have a PR to fix it